### PR TITLE
Cancel broadcast via API

### PR DESCRIPTION
**Pull request by CrystalPea**
Allow broadcast to be cancelled via an API request from a service.

Should go in after: https://github.com/alphagov/notifications-admin/pull/4135
 
Part of this ticket: https://www.pivotaltracker.com/story/show/176927105
To do in a separate PR: audit which API key cancelled the broadcast

### CODE REVIEW

**Review by benthorner — COMMENTED** [review]
Tested locally and works as expected ✅.

Nice one keeping this secure by scoping the broadcast query to the authenticated service 💯. I also found the tests easy to follow, although I think we may be missing one or two.

There are a couple of parts of the main code that I'm not sure about, so I won't approve just yet but this is pretty much there in doing what it's meant to do.

File: .python-version [hunk 1]
```diff
@@ -0,0 +1 @@
+3.9.7
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply .python-version:1]: Ahh, good find, I will get rid of it now and add the config that you recommended to my local

[the remark this answers is not in the document]
Comment (CrystalPea) [reply .python-version]: That's done 👍🏼

File: app/v2/broadcast/post_broadcast.py [hunk 2]
```diff
@@ -3,9 +3,16 @@
 from flask import current_app, jsonify, request
 from notifications_utils.polygons import Polygons
 from notifications_utils.template import BroadcastMessageTemplate
+from sqlalchemy.orm.exc import NoResultFound
 
 from app import api_user, authenticated_service
+from app.broadcast_message.rest import (
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/v2/broadcast/post_broadcast.py:9]: That is a good point, I did think about moving it out, was not sure what the best place would be, dao file or a new utils file in broadcast_message, and worried we would end up duplicating tests even more, so ended up leaving it where it was.

Still, you are right we don't usually import from rest and could have tests more at source. I will move it to a utils file I think.

File: app/v2/broadcast/post_broadcast.py [hunk 3]
```diff
@@ -37,52 +44,85 @@ def create_broadcast():
...
+        broadcast_message = dao_get_broadcast_message_by_references_and_service_id(
+            references_to_original_broadcast,
+            service_id
+        )
+    except NoResultFound:
+        raise BadRequestError(
+            message="Broadcast message reference and service id didn't match with any existing broadcasts",
+            status_code=404,
+        )
+    # do we need to check if service is active?
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/v2/broadcast/post_broadcast.py:114]: it was a TODO to check if it's needed that I forgot about, thanks for answering it, I can remove it now :D

File: app/v2/broadcast/post_broadcast.py [hunk 4]
```diff
@@ -37,52 +44,85 @@ def create_broadcast():
...
 
-    return jsonify(broadcast_message.serialize()), 201
+
+def _cancel_or_reject_broadcast(references_to_original_broadcast, service_id):
+    try:
+        broadcast_message = dao_get_broadcast_message_by_references_and_service_id(
+            references_to_original_broadcast,
+            service_id
+        )
+    except NoResultFound:
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/v2/broadcast/post_broadcast.py:109]: Hmmm I wanted to give a clear message to the users, but maybe you are right that the context is enough.

File: app/broadcast_message/rest.py [hunk 5]
```diff
@@ -42,10 +42,10 @@ def _parse_nullable_datetime(dt):
     return dt
 
 
-def _update_broadcast_message(broadcast_message, new_status, updating_user):
+def validate_and_update_broadcast_message_status(broadcast_message, new_status, updating_user, from_api=False):
     if updating_user not in broadcast_message.service.users:
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/broadcast_message/rest.py:46]: Cool I've done it now, will push changes in a mo

File: tests/app/v2/broadcast/test_post_broadcast.py [hunk 6]
```diff
@@ -119,6 +119,126 @@ def test_valid_post_cap_xml_broadcast_returns_201(
...
+
+    response_for_cancel = client.post(
+        path='/v2/broadcast',
+        data=sample_cap_xml_documents.WAINFLEET_CANCEL,
+        headers=[('Content-Type', 'application/cap+xml'), auth_header],
+    )
+    assert response_for_cancel.status_code == 201
+
+    assert len(broadcast_message.events) == 1
+    alert_event = broadcast_message.events[0]
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/v2/broadcast/test_post_broadcast.py:199]: I think since I plan to move this part of the code to a utils file and test it separately, we will not need extra test for this in the API flow (since it will be the common util part that deals with this, and it will be tested there, at source)

File: app/v2/broadcast/post_broadcast.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/v2/broadcast/post_broadcast.py]: Ah, I wasn't sure if it was a conscious decision. I don't feel too strongly about it - it's a subjective call about which error message may be more helpful. Thanks for changing it, though 👍.

File: tests/app/v2/broadcast/test_post_broadcast.py [hunk 7]
```diff
@@ -119,6 +119,126 @@ def test_valid_post_cap_xml_broadcast_returns_201(
...
+    mock_task.assert_called_once_with(kwargs={'broadcast_event_id': str(alert_event.id)}, queue='broadcast-tasks')
+
+    assert broadcast_message.status == 'cancelled'
+    assert broadcast_message.cancelled_at is not None
+    assert broadcast_message.cancelled_by_id is None  # broadcast cancelled via API, so not by any single user
+    assert broadcast_message.updated_at is not None
+
+
+def test_cancel_request_does_not_cancel_broadcast_if_reference_does_not_match(
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/app/v2/broadcast/test_post_broadcast.py]: ok, done 👍🏼

File: app/v2/broadcast/post_broadcast.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CrystalPea) [reply app/v2/broadcast/post_broadcast.py]: Ok, this has been moved to utils now, and test have been moved and adjusted.

[the remark this answers is not in the document]
Comment (benthorner) [reply app/v2/broadcast/post_broadcast.py]: Great stuff - that's a lot of tests moved 🏅. I like the balance we now have: one file with a few tests for the actual requests + another file with lots of other tests for edge cases and details. Calling the function directly is also simpler than figuring out what request to make, so this will make it easier to extend in future.

**Review by benthorner — APPROVED** [review]
Everything looks good. 

Thanks for doing the extra work to refactor the functions ⭐.
